cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java

```diff
@@ -103,50 +103,46 @@ public boolean optimize(ClusterModel clusterModel, Set<Goal> optimizedGoals, Opt
-                         + " for leadership {}.", p.topicPartition(), leaderCandidate, excludedBrokersForLeadership);
-                continue;
-              }
-              clusterModel.relocateLeadership(r.topicPartition(), p.leader().broker().id(), leaderCandidate.id());
-              relocatedLeadership = true;
-            }
-            if (clusterModel.demotedBrokers().contains(leaderCandidate)) {
-              LOG.warn("The leader of partition {} has to be on a demoted broker {} because all the alive "
-                       + "replicas are demoted.", p.topicPartition(), leaderCandidate.id());
-            }
-            if (r.disk() != null && r.disk().state() == Disk.State.DEMOTED) {
-              LOG.warn("The leader of partition {} has to be on a demoted disk {} of broker {} because all the alive "
-                       + "replicas are demoted.", p.topicPartition(), r.disk().logDir(), leaderCandidate.id());
-            }
-            break;
+        Replica candidateReplica = p.replicas().get(0);
+        Broker candidateBroker = candidateReplica.broker();
+        if (!candidateReplica.isLeader()) {
+          if (candidateReplica.isCurrentOffline()) {
+            sb.append(String.format("The preferred replica of partition %s on broker %s is offline.%n", p.topicPartition(), candidateBroker));
+            failedPartitions++;
+          } else if (excludedBrokersForLeadership.contains(candidateBroker.id())) {
+            sb.append(String.format("Skipped leadership transfer of partition %s to broker %d because it is among brokers excluded"
+                                    + " for leadership %s.%n", p.topicPartition(), candidateBroker.id(), excludedBrokersForLeadership));
+            failedPartitions++;
+          } else if (clusterModel.demotedBrokers().contains(candidateBroker)) {
+            sb.append(String.format("Skipped leadership transfer of partition %s to broker %d because it is among recently "
+                                    + "demoted brokers %s.%n\" ", p.topicPartition(), candidateBroker.id(), clusterModel.demotedBrokers()));
+            failedPartitions++;
+          } else if (candidateReplica.disk() != null && candidateReplica.disk().state() == Disk.State.DEMOTED) {
+            sb.append(String.format("Skipped leadership transfer of partition %s to broker %d because the hosting disk %s is recently demoted.%n",
+                                    p.topicPartition(), candidateBroker.id(), candidateReplica.disk().logDir()));
+            failedPartitions++;
+          } else {
+            clusterModel.relocateLeadership(candidateReplica.topicPartition(), p.leader().broker().id(), candidateBroker.id());
           }
         }
       }
     }
+    if (failedPartitions > 0) {
+      LOG.warn("Leadership transfer of {} partitions are skipped. Details:\n{}", failedPartitions, sb.toString());
```

Comment:
Nit: Using `\n` makes it harder to parse and print `warn` logs -- can we avoid it?